File: src/main/javabak/item/WrenchMode.java
package redcrafter07.processed.item;

import io.netty.buffer.ByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.util.ByIdMap;
import net.minecraft.util.StringRepresentable;

import java.util.function.IntFunction;

public enum WrenchMode implements StringRepresentable {
    Config(0, "config"), Rotate(1, "rotate");

    WrenchMode(int id, String modeName) {
        this.id = id;
        this.modeName = modeName;
    }

    private final int id;
    private final String modeName;

    public final static IntFunction<WrenchMode> BY_ID = ByIdMap.continuous(WrenchMode::getId, values(), ByIdMap.OutOfBoundsStrategy.WRAP);
    public final static EnumCodec<WrenchMode> CODEC = StringRepresentable.fromEnum(WrenchMode::values);
    public final static StreamCodec<ByteBuf, WrenchMode> STREAM_CODEC = ByteBufCodecs.idMapper(BY_ID, WrenchMode::getId);

    @Override
    public String getSerializedName() {
        return modeName;
    }

    public int getId() {
        return id;
    }

    public WrenchMode next() {
        return BY_ID.apply(getId() + 1);
    }

    public WrenchMode previous() {
        return BY_ID.apply(getId() - 1);
    }

    public String translation() {
        return "item.processed.wrench.mode." + modeName;
    }
}

File: src/main/kotlin/redcrafter07/processed/items/WrenchMode.kt
package redcrafter07.processed.items

import com.mojang.serialization.Codec
import io.netty.buffer.ByteBuf
import net.minecraft.network.chat.Component
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.util.ByIdMap
import net.minecraft.util.StringRepresentable
import net.neoforged.neoforge.common.TranslatableEnum
import redcrafter07.processed.Translations
import java.util.function.IntFunction

enum class WrenchMode(val id: Int, val modeName: String) : StringRepresentable, TranslatableEnum {
    Config(0, "config"), Rotate(1, "rotate");

    companion object {
        val BY_ID: IntFunction<WrenchMode> =
            ByIdMap.continuous(WrenchMode::id, WrenchMode.entries.toTypedArray(), ByIdMap.OutOfBoundsStrategy.WRAP)
        val CODEC: Codec<WrenchMode> = StringRepresentable.fromValues(WrenchMode::values)
        val STREAM_CODEC: StreamCodec<ByteBuf, WrenchMode> = ByteBufCodecs.idMapper(BY_ID, WrenchMode::id)
    }

    val next: WrenchMode get() = BY_ID.apply(id + 1)
    val previous: WrenchMode get() = BY_ID.apply(id - 1)

    override fun getSerializedName(): String = modeName
    override fun getTranslatedName(): Component = Translations.wrenchMode(modeName)
}